include/seastar/core/semaphore.hh

```diff
@@ -214,6 +221,28 @@ public:
     {
         static_assert(std::is_nothrow_move_constructible_v<expiry_handler>);
     }
+
+    basic_semaphore(const basic_semaphore&) = delete;
+    basic_semaphore(basic_semaphore&& o) noexcept
+        : _count(std::exchange(o._count, 0))
+        , _outstanding_units(0)
+        , _ex(std::move(o._ex))
+        , _wait_list(std::move(o._wait_list))
+    {
+        assert(o._outstanding_units == 0 && "Cannot move a semaphore with outstanding semaphore units");
+    }
```

Comment:
Let's just deprecate the move constructor. It's not safe.

People who want to move an object containing a semaphore can initialize the new one from scratch.

Reply:
It's not that simple, we need to move it when it's freshly constructed for example in `do_with`

Reply:
also returning it as a future result also requires it will be (nothrow) move constructible.

Reply:
Yes